src/python/pants/backend/jvm/rules/targets.py

```diff
@@ -0,0 +1,808 @@
+
+    By default, this uses the `name` field. Note this default is unsafe because of the possible
+    conflict when multiple binaries are built.
+    """
+
+    alias = "basename"
+
+
+class JvmBinaryDeployExcludes(JvmExcludes):
+    """A list of `exclude` objects to apply at deploy time.
+
+    If you, for example, deploy a Java servlet that has one version of `servlet.jar` onto a Tomcat
+    environment that provides another version, they might conflict. `deploy_excludes` gives you a
+    way to build your code but exclude the conflicting `jar` when deploying.
+    """
+
+    alias = "deploy_excludes"
+
+
+class JvmBinaryDeployJarRules(PrimitiveField):
+    """A `jar_rules` object for packaging this binary in a deploy JAR."""
+
+    alias = "deploy_jar_rules"
+    value: Optional[JarRules]
+    default = None
+
+    @classmethod
+    def compute_value(
+        cls, raw_value: Optional[JarRules], *, address: Address
+    ) -> Optional[JarRules]:
+        value_or_default = super().compute_value(raw_value, address=address)
+        if value_or_default is not None and not isinstance(value_or_default, JarRules):
+            raise InvalidFieldTypeException(
+                address, cls.alias, value_or_default, expected_type="a `jar_rules` object"
+            )
+        return value_or_default
+
+
+class JvmBinaryManifestEntries(DictStringToStringField):
+    """A dict that specifies entries for `ManifestEntries` for adding to MANIFEST.MF when packaging
+    this binary."""
+
+    alias = "manifset_entries"
+
+
+class JvmBinaryShadingRules(PrimitiveField):
+    """A list of shading rules to apply when building a shaded (aka monolithic aka fat) binary jar.
+
+    The order of the rules matters: the first rule which matches a fully-qualified class name is
+    used to shade it. See shading_relocate(), shading_exclude(), shading_relocate_package(), and
+    shading_exclude_package().
+    """
+
+    alias = "shading_rules"
+    value: Tuple[Any, ...]
+    default = None
+
+    @classmethod
+    def compute_value(
+        cls, raw_value: Optional[Iterable[Any]], *, address: Address
+    ) -> Tuple[Any, ...]:
+        value_or_default = super().compute_value(raw_value, address=address)
+        return tuple(value_or_default)
+
+
+class JvmBinary(Target):
+    """A JVM binary."""
+
+    alias = "jvm_binary"
+    core_fields = (
+        *COMMON_JVM_FIELDS,
+        Sources,
+        ExtraJvmOptions,
+        JvmRuntimePlatform,
+        JvmBinaryMain,
+        JvmBinaryBasename,
+        JvmBinaryDeployExcludes,
+        JvmBinaryDeployJarRules,
+        JvmBinaryManifestEntries,
+        JvmBinaryShadingRules,
+    )
+
+
+# -----------------------------------------------------------------------------------------------
+# `jvm_prep_command` target
+# -----------------------------------------------------------------------------------------------
+
+
+class JvmPrepCommandMainClass(StringField):
```

Comment:
Ughhhh... We should really just be using `JvmBinaryMain` here, but they have different aliases! :(

(Similarly for jvm options)

Reply:
Indeed :/ we could directly subclass those and simply change the alias, though! Then we can do things like leverage the docstring from `JvmBinaryMain`.

I'll fix this.